Source/WebCore/platform/network/mac/ResourceErrorMac.mm

```diff
@@ -28,22 +28,33 @@
 
 #import <CoreFoundation/CFError.h>
 #import <Foundation/Foundation.h>
+#import <pal/spi/cocoa/NWSPI.h>
 #import <wtf/BlockObjCExceptions.h>
 #import <wtf/NeverDestroyed.h>
 #import <wtf/URL.h>
 #import <wtf/text/WTFString.h>
 
-static bool failureIsDueToTrackerBlocking(NSError *error)
+static String extractBlockedTrackerHostName(NSError *error)
 {
 #if ENABLE(ADVANCED_PRIVACY_PROTECTIONS)
     for (NSError *underlyingError in error.underlyingErrors) {
-        if ([[underlyingError.userInfo objectForKey:@"_NSURLErrorBlockedTrackerFailureKey"] boolValue])
-            return true;
+        auto userInfo = underlyingError.userInfo;
+        if ([userInfo[@"_NSURLErrorBlockedTrackerFailureKey"] boolValue]) {
+            String blockedTrackerName;
+            if (id failingPath = userInfo[@"_NSURLErrorNWPathKey"]) {
+                auto failingEndpoint = adoptNS(nw_path_copy_effective_remote_endpoint(failingPath));
+                if (auto* hostName = nw_endpoint_get_known_tracker_name(failingEndpoint.get()))
+                    blockedTrackerName = String::fromUTF8(hostName);
+                else
+                    blockedTrackerName = emptyString();
```

Comment:
Minor nit: Seems like you could drop this, return early with the value from `String::fromUTF8`, and then use a String::isNull check in NetworkResourceLoader - because it doesn't seem like the empty string give more value than a null string. But feel free to ignore this suggestion.

Reply:
The null string in this case indicates that nothing was blocked — the empty string is just a bit of defensive programming against the case where `nw_endpoint_get_known_tracker_name` returns null, but perhaps this is enough of an edge case that we can just assume the tracker name is always known.

Reply:
That distinction could be useful, but the current isEmpty() check in `NetworkResourceLoader::didFailLoading` combines those two cases ("nothing was blocked" and "get_known_tracker_name returns null"), so I didn't see much value in returning emptyString() vs. a null String. I see your point about handling the case where "get_known_tracker_name returns null" because that is clearly different from "nothing was blocked", but this patch falls back on sending the unmodified URL from `error.failingURL()` in the DidBlockLoadToKnownTracker message, and that's just the current behavior.

Reply:
Oh, my original patch below used `isNull()` to determine whether a tracker was blocked, and `isEmpty()` to determine whether a tracker host name is known (and in doing so, sneakily allowed for the case where `get_known_tracker_name` returns null to be handled gracefully without treating it as a non-blocked tracker).

I've since adjusted this logic to be more clear by explicitly falling back on the failing URL's host (it should result in the same behavior, only easier to follow).